### PR TITLE
MAINT: Add placeholder stubs for `ndarray` and `generic`

### PR DESCRIPTION
**Pull request by BvB93**
This pull requests removes the class-level `__getattr__` method from the `_ArrayOrScalarCommon` 
super-class, adding placeholder stubs for the previously missing methods. In other words: all 
methods in `np.ndarray` and `np.generic` are now explicitly defined.

Examples
---------
Previously this would pass as `__getattr__` was annotated as returing `Any`, _i.e._ any `getattr()` 
operation (and any protocol by extension) was considered legal.
``` python
from typing import Iterator
import numpy as np

A: Iterator = np.array(0)  # This would previously pass
```

### CODE REVIEW

File: numpy/__init__.pyi [hunk 1]
```diff
@@ -293,84 +399,20 @@ class ndarray(_ArrayOrScalarCommon, Iterable, Sized, Container):
...
     # Many of these special methods are irrelevant currently, since protocols
     # aren't supported yet. That said, I'm adding them for completeness.
     # https://docs.python.org/3/reference/datamodel.html
     def __len__(self) -> int: ...
-    def __getitem__(self, key) -> Any: ...
     def __setitem__(self, key, value): ...
     def __iter__(self) -> Any: ...
     def __contains__(self, key) -> bool: ...
     def __index__(self) -> int: ...
+    def __matmul__(self, other): ...
```
Comment (BvB93) [inline numpy/__init__.pyi:413]: `__matmul__` and the likes is only defined for `np.ndarray`; not `np.generic`

File: numpy/__init__.pyi [hunk 2]
```diff
@@ -293,84 +399,20 @@ class ndarray(_ArrayOrScalarCommon, Iterable, Sized, Container):
     @shape.setter
     def shape(self, value: _ShapeLike): ...
     @property
-    def flat(self: _ArraySelf) -> flatiter[_ArraySelf]: ...
-    @property
     def strides(self) -> _Shape: ...
     @strides.setter
     def strides(self, value: _ShapeLike): ...
-    # Array conversion
```
Comment (BvB93) [inline numpy/__init__.pyi:301]: These methods are present in both `np.ndarray` and `np.generic`.
Note that, as the `_ArraySelf` typevar is bound to the `_ArrayOrScalarCommon` superclass, these methods can be moved without further adjustments.

File: numpy/__init__.pyi (hunk 1 above)
Comment (eric-wieser) [reply numpy/__init__.pyi:413]: If you're doing this type of cleanup... All the `__i<op>` methods are also present only on arrays.

File: numpy/__init__.pyi [hunk 3]
```diff
@@ -253,15 +253,121 @@ class _ArrayOrScalarCommon(
...
+    def put(self, indices, values, mode=...): ...
+    def repeat(self, repeats, axis=...): ...
+    def round(self, decimals=..., out=...): ...
+    def searchsorted(self, v, side=..., sorter=...): ...
+    def setfield(self, val, dtype, offset=...): ...
+    def sort(self, axis=..., kind=..., order=...): ...
+    def std(self, axis=..., dtype=..., out=..., ddof=..., keepdims=...): ...
+    def sum(self, axis=..., dtype=..., out=..., keepdims=..., initial=..., where=...): ...
+    def take(self, indices, axis=..., out=..., mode=...): ...
+    def tostring(self, order=...): ...
```
Comment (eric-wieser) [inline numpy/__init__.pyi:368]: If it's relevant, `tostring` is deprecated to we could perhaps skip it

Comment (person142) [reply numpy/__init__.pyi:368]: Yeah we have in fact removed it from the subs before...

Comment (BvB93) [reply numpy/__init__.pyi:368]: I'll add a note in a bit as a future reminder.

File: numpy/__init__.pyi (hunk 1 above)
Comment (BvB93) [reply numpy/__init__.pyi:392]: Might as well while we're at it; done as of https://github.com/numpy/numpy/pull/17188/commits/478c3f4ba320e580673da8a97b57ef45d96dac4a.

File: numpy/__init__.pyi (hunk 3 above)
Comment (BvB93) [reply numpy/__init__.pyi]: And it's been removed (again) as of https://github.com/numpy/numpy/pull/17188/commits/5ea2ec96779cb529776487b26a261ad08a43c74a.

**Comment by BvB93** [comment]
Rebased to get rid the merge conflict introduced in https://github.com/numpy/numpy/pull/17235 (thanks for the headsup @charris).

**Comment by charris** [comment]
@BvB93 I'm inclined to just merge all the typing PRs and then move the tests into `nump/typing`. The work is ongoing, but at least that would provide a base to work from. @rgommers @person142 @mattip thoughts?

**Comment by mattip** [comment]
I don't know enough to have an opinion.

**Comment by BvB93** [comment]
I'd hold of on merging https://github.com/numpy/numpy/pull/17117 as that PR could first use a bit more discussion.
The other four are ready to go, assuming there are no further comments.

**Comment by charris** [comment]
OK, I'll do the merges and moves. The remaining drafts and #17117 will just need to deal with that.

**Comment by charris** [comment]
Needs another rebase.

**Comment by charris** [comment]
I fixed the merge conflict.

**Comment by charris** [comment]
@BvB93 There are now 3 testing errors, probably due to the preceding PRs being merged.

**Comment by BvB93** [comment]
Working on it; expect a push within a few minutes.

**Comment by charris** [comment]
Great. There are also two draft PRs, #17096 and #16759.  Do they need to be kept open?

**Comment by BvB93** [comment]
> Great. There are also two draft PRs, #17096 and #16759. Do they need to be kept open?

Let's keep them; they both contain some useful additions but it's simply too early to merge them.

**Comment by BvB93** [comment]
Found the issue btw (https://github.com/numpy/numpy/pull/17188/commits/599b5af0bc62e0e2be188e03eaeef357b901bf93).

**Comment by charris** [comment]
> Let's keep them

OK. I'm going to move the tests after this PR goes in, so they will need revision. I'm don't know what to do about the random problem if/when we get around to annotating that module.

**Comment by BvB93** [comment]
> OK. I'm going to move the tests after this PR goes in, so they will need revision. I'm don't know what to do about the random problem if/when we get around to annotating that module.

Merging https://github.com/numpy/numpy/pull/17104 seems to have fixed the most immediate issue.
The objects within the module are still annotated as `Any`, but at least it can now differentiate between objects which are and aren't supposed to be in the module.

``` python
from typing import Type
import numpy as np

reveal_type(np.random)  # note: Revealed type is '_importlib_modulespec.ModuleType'
reveal_type(np.random.rand)  # note: Revealed type is 'Any'
reveal_type(np.random.bogus_name)  # error: Module has no attribute "bogus_name"

```

**Comment by charris** [comment]
In it goes, not waiting for shippable. Thanks @BvB93 .

**Pull request merged**